### PR TITLE
GSOC17 - Facemark API

### PR DESCRIPTION
**Pull request by kurnianggoro**
**Merge with extra**: https://github.com/opencv/opencv_extra/pull/365

Project: API for Facial Landmark Detector
Mentor: Delia Passalacqua

This pull request contains the implementation of Facemark API as well as two facemark algorithm, AAM and LBF.

video: https://www.youtube.com/watch?v=B7WGyhl2zm8
report: https://gist.github.com/kurnianggoro/74de9121e122ad0bd825176751d47ecc

Currently, there are several interfaces provided in the Facemark base class including:
- load training data
- set custom face detector
- face detection
- training the model
- saving a trained model
- loading a trained model
- fitting

For the AAM algorithm, all of its training functionalities are already coded.
Several related functionalities are listed as follows:
 - Procrustes analysis
 - PCA projection
 - customized Delaunay triangulation
 - warp image
 - Gram-Schmidth orthonormalization
 - feature extraction

### CODE REVIEW

**Comment by sergboec** [comment]
@kurnianggoro  Do you have plans to add implementation of algorithm from dlib ?

**Comment by alalek** [comment]
@kurnianggoro What is the license of images from opencv_extra patch?

**Comment by kurnianggoro** [comment]
hi @alalek , I just found them on google, will it be problem?

**Comment by alalek** [comment]
test_stereo has sporadic failures - build restarted.

All images should have known proper usage rights. It is better to reuse existed OpenCV images, or use Google [filtering](https://support.google.com/websearch/answer/29508?hl=en) (but anyway double check is required).

**Comment by kurnianggoro** [comment]
@alalek hi, i have changed the file and provide the source links (the license is available,you can check), they are from wikipedia, so i hope it will not be problem

**Comment by vpisarev** [comment]
@kurnianggoro, let's not put faces of public people (or not public, it does not matter) to OpenCV. lena is the exception, 'cause it's considered a standard test image.

File: modules/face/src/facemarkLBF.cpp [hunk 1]
```diff
@@ -0,0 +1,1294 @@
...
+        if(params.verbose) printf("loading data from : %s\n", s.c_str());
+        std::ifstream infile;
+        infile.open(s.c_str(), std::ios::in);
+        if (!infile) {
+           std::string error_message = "No valid input file was given, please check the given filename.";
+           CV_Error(CV_StsBadArg, error_message);
+        }
+
+        FILE *fd = fopen(s.c_str(), "rb");
+        regressor.read(fd, params);
```
Comment (alalek) [inline modules/face/src/facemarkLBF.cpp]: OpenCV modules should not read/write files directly, especially via old "C-API" (there are a lot of missing error checks at least; there is no in-memory storage support; cross-platform issues).
Why FileStorage doesn't work here?

File: modules/face/src/facemarkLBF.cpp [hunk 2]
```diff
@@ -0,0 +1,1294 @@
...
+        std::random_shuffle(bboxes.begin(), bboxes.end());
+        std::srand(seed);
+        std::random_shuffle(current_shapes.begin(), current_shapes.end());
+
+
+        regressor.initRegressor(params);
+        regressor.trainRegressor(imgs, gt_shapes, current_shapes, bboxes, mean_shape, 0, params);
+
+        FILE *fd = fopen(params.model_filename.c_str(), "wb");
+        assert(fd);
```
Comment (alalek) [inline modules/face/src/facemarkLBF.cpp]: `assert()` should not be used in OpenCV, use `CV_Assert()` instead.

File: modules/face/src/facemarkLBF.cpp [hunk 3]
```diff
@@ -0,0 +1,1294 @@
...
+                imgs[idx] = data_faces[i];
+                gt_shapes[idx] = data_shapes[i];
+                bboxes[idx] = data_boxes[i];
+                current_shapes[idx] = data_boxes[i].reproject(data_boxes[k].project(data_shapes[k]));
+            }
+        }
+
+        // random shuffle
+        unsigned int seed = (unsigned int)std::time(0);
+        std::srand(seed);
```
Comment (alalek) [inline modules/face/src/facemarkLBF.cpp]: `rand()` is not thread-safe - can't provide reproducible results (for testing).
Consider to use cv::rand() , cv::theRNG() instead.

File: modules/face/src/liblinear.hpp [hunk 4]
```diff
@@ -0,0 +1,3968 @@
...
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#ifdef F2C_COMPAT
+
+void cdotc_(fcomplex *dotval, int *n, fcomplex *cx, int *incx,
+            fcomplex *cy, int *incy);
+
+void cdotu_(fcomplex *dotval, int *n, fcomplex *cx, int *incx,
```
Comment (alalek) [inline modules/face/src/liblinear.hpp]: Using of external libraries should be avoided. Especially in case of duplicated functionality.
Why does `cv::ml::LogisticRegression` not work for you?

File: modules/face/test/test_facemark_lbf.cpp [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+    }
+
+    EXPECT_NO_THROW(facemark->training());
+}
+
+TEST(CV_Face_FacemarkLBF, can_load_model) {
+    string model_filename = "LBF.model";
+
+    Ptr<Facemark> facemark = FacemarkLBF::create();
+    EXPECT_NO_THROW(facemark->loadModel(model_filename.c_str()));
```
Comment (alalek) [inline modules/face/test/test_facemark_lbf.cpp]: Each test should be standalone and don't rely on results of other tests.
Also tests should not write any data (models) if it is not required.

Probably we should put model somewhere or we can merge train/load tests into one.

**Comment by kurnianggoro** [comment]
@vpisarev the test function needs at least 2 images, so i put mark zuckerberg and downey johnson photos from wikipedia (the license allows it, CCA 2.0 and CCA 3.0), will that be a problem? also regarding to the lena image, in other library it was removed due to copyright matter, i am not sure why opencv still have it. check it here https://github.com/torch/nn/issues/854 also in wikipedia it is stated that lena is non free image https://en.wikipedia.org/wiki/File:Lenna.png

File: modules/face/src/facemarkLBF.cpp (hunk 3 above)
Comment (kurnianggoro) [reply modules/face/src/facemarkLBF.cpp]: according to the documentation http://docs.opencv.org/3.2.0/d2/de8/group__core__array.html#ga75843061d150ad6564b5447e38e57722 theRNG function will use different random generator for each thread, but in this case the program only needs to randomize the order of training sample, which should produce same random order for each sample category (i.e. the randomized order or image sample should corresponds to the randomized order of the annotation). If reproducible results is expected, then i think we just need to set the seed value to a fixed number, or just remove this shuffling part (i don't think it will produce significantly different results)

File: modules/face/src/facemarkLBF.cpp (hunk 1 above)
Comment (kurnianggoro) [reply modules/face/src/facemarkLBF.cpp]: ok, i will try to change to filestorage.

File: modules/face/src/liblinear.hpp (hunk 4 above)
Comment (kurnianggoro) [reply modules/face/src/liblinear.hpp]: I tried to remove the liblinear dependency but was not successful (using gradient descent regression and simple regression w^T x = y). The LogisticRegression result is in range 0~1, isn't it? I considered to use this function before but worried that the result will be different to the original version (i need to normalize the target values, i.e. the landmark points, into 0~1 range,maybe by dividing the values by width and height of input image, and not sure about the result) and the deadline was approaching, so i did not test LogisticRegression. But let me try now...

File: modules/face/test/test_facemark_lbf.cpp (hunk 5 above)
Comment (kurnianggoro) [reply modules/face/test/test_facemark_lbf.cpp]: merging the train-load test is easier, i can put a flag for not saving the model.

File: modules/face/src/facemarkLBF.cpp [hunk 6]
```diff
@@ -0,0 +1,1294 @@
...
+
+        cascade_face = "../data/haarcascade_frontalface_alt.xml";
+        shape_offset = 0.0;
+        n_landmarks = 68;
+        initShape_n = 10;
+        stages_n=5;
+        tree_n=6;
+        tree_depth=5;
+        bagging_overlap = 0.4;
+        model_filename = "LBF.model";
```
[the remark this answers is not in the document]
Comment (kurnianggoro) [reply modules/face/src/facemarkLBF.cpp]: this is just the default filename for the saving file, is it also problem? if it were for reading file like the cascade face then for sure it will be problem

File: modules/face/src/facemarkLBF.cpp (hunk 3 above)
Comment (alalek) [reply modules/face/src/facemarkLBF.cpp]: `cv::theRNG` is thread-safe (maintains own state for each thread).
Seed initialization from `time()` should be removed from algorithm implementations - seed should be controlled externally.

Regarding this exact algorithm, I believe it should work without shuffling of input data.

File: modules/face/src/facemarkLBF.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (alalek) [reply modules/face/src/facemarkLBF.cpp]: This is minor problem (at least there is no relative path to subdirectory).

File: modules/face/src/facemarkLBF.cpp (hunk 1 above)
Comment (alalek) [reply modules/face/src/facemarkLBF.cpp]: Well, there is similar PR #1199 with same [problem](https://github.com/opencv/opencv_contrib/pull/1199/files#diff-bda6c2019eab834caf1e9453102e8025). Probably we should keep this as is if there is some problem with `cv::FileStorage`.
I believe it is fine for the first version. So consider this problem as minor too.

File: modules/face/src/liblinear.hpp (hunk 4 above)
Comment (alalek) [reply modules/face/src/liblinear.hpp]: OK. Thank you for information.

**Comment by alalek** [comment]
AAM tests are crashed in the "Debug" [build](http://pullrequest.opencv.org/buildbot/builders/precommit-contrib_linux64_no_opt/builds/10037). Please take a look.

File: modules/face/src/facemarkAAM.cpp [hunk 7]
```diff
@@ -560,8 +560,8 @@ namespace face {
         reduce(Ys,sumYs, 0, CV_REDUCE_SUM);
 
         //calculate the normrnd
-        double normX = sqrt(sumXs.at<float>(0)+sumXs.at<float>(1));
-        double normY = sqrt(sumYs.at<float>(0)+sumYs.at<float>(1));
+        double normX = sqrt(Mat(sumXs.reshape(1)).at<float>(0)+Mat(sumXs.reshape(1)).at<float>(1));
```
Comment (kurnianggoro) [inline modules/face/src/facemarkAAM.cpp]: @alalek the error in build debug is fixed here, but why it is not checked anymore?

Comment (alalek) [reply modules/face/src/facemarkAAM.cpp]: This build is optional, it is not triggered automatically.
Debug build is passed, thanks!

File: modules/face/test/test_facemark_aam.cpp [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        image = imread(images_train[i].c_str());
+        EXPECT_TRUE(loadFacePoints(points_train[i].c_str(),landmarks));
+        EXPECT_TRUE(landmarks.size()>0);
+        EXPECT_TRUE(facemark->addTrainingSample(image, landmarks));
+    }
+
+    EXPECT_NO_THROW(facemark->training());
+}
+
+TEST(CV_Face_FacemarkAAM, can_load_model) {
```
Comment (alalek) [inline modules/face/test/test_facemark_aam.cpp]: The same note about standalone tests.

**Comment by kurnianggoro** [comment]
Hi @alalek , I already fixed all the problems you've mentioned, but the building bot failed due to stereo module, can you please restart the building? also regarding the liblinear depency, I've tried to use opencv's logistric regression and neural network but both of them are not compatible. For the log-res, it seems like that the algorithm automatically assign categories to the training data and the number of resulted categories are inconsistent across different training data. As for the neural network, it do not support linear activation, I tried to train but the results are nan. So, I grab the support vector regression code from the liblinear, it just one function, not too long. I've tested it and the result is good. You can try to train using the Helen dataset if you want. The tutorial is here http://pullrequest.opencv.org/buildbot/export/pr_contrib/1257/docs/d7/dec/tutorial_facemark_usage.html but make sure your system has around 9GB of RAM and just wait 45in~1hour for the for the training. I've tried to train on smaller dataset but the results are not good.

**Comment by alalek** [comment]
@kurnianggoro Thank you! Update looks good to me.

**Comment by alalek** [comment]
The current issue is related to test images. We can't accept such risk.
Could you please replace these images? For example, from free datasets (but the most part of datasets is not free/restricted/missing license), or very old photos.

File: modules/face/src/facemarkLBF.cpp [hunk 9]
```diff
@@ -1191,6 +1181,185 @@ namespace face {
...
+            }
+            index[i] = i;
+        }
+
+        while(iter < max_iter){
+            Gmax_new = 0;
+            Gnorm1_new = 0;
+
+            for(i=0; i<active_size; i++){
+                int j = i+rand()%(active_size-i);
```
Comment (alalek) [inline modules/face/src/facemarkLBF.cpp]: Please avoid using of `rand()` - it doesn't provide reproducible results.

Like "stereo" test (it uses `rand()` too), which fails time-to-time.

**Comment by kurnianggoro** [comment]
hi @alalek i have changed the test images using frames taken from the tracking module test video, I hope it will not be a problem now.

**Comment by alalek** [comment]
Thank you! Looks good to me :+1:

**Comment by Macbull** [comment]
Is there a python wrapper of this API in any release version?
Thanks
